server-spi-private/src/main/java/org/keycloak/models/LockObjectsForModification.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models;
+
+/**
+ * This flags the session that all information loaded from the stores should be locked as the service layer
+ * plans to modify it.
+ *
+ * This is just a hint to the underlying storage, and a store might choose to ignore it.
+ * The lock for any object retrieved from the session will be kept until the end of the transaction.
+ *
+ * If the store supports it, this could prevent exceptions due to optimistic locking
+ * problems later in the processing. If the caller retrieved objects without this wrapper, they would still be
+ * able to modify those objects, and those changes would be written to the store at the end of the transaction at the lastet,
+ * but they won't be locked.
+ *
+ *
+ * @author Alexander Schwartz
+ */
+public class LockObjectsForModification {
+
+    private static final String ATTRIBUTE = LockObjectsForModification.class.getCanonicalName();
+
+    public static LockObjectsForModification.Enabled enable(KeycloakSession session) {
+        return new Enabled(session);
+    }
+
+    public static boolean isEnabled(KeycloakSession session) {
+        return session.getAttribute(ATTRIBUTE) != null;
+    }
+
+    public static <V> V lockObjectsForModification(KeycloakSession session, CallableWithoutThrowingAnException<V> callable) {
```

Comment:
After creating this method, I found that this would lock all entities retrieved from the DB in the Callable. The current way to avoid locking Clients is that the transaction has a flag lockingSupportedForEntity().

Thinking about this a bit more, the call could be more specific here: Instead of lockObjectsForModification it could be lockUsersSessionsForModification. This would provide a clearer intent what the caller tries to achieve. Instead of setting a boolean flag, it would then populate a Map with the models to be locks (UserSessionModel.class in this case).

I'd like to hear your thoughts. I'd suggest a new issue for that. It should still be part of the next upcoming release to avoid spreading this API.

Reply:
I agree with that idea.

To elaborate on specifics, would you store in the `KeycloakSession` attribute a `Set` of `Class` objects that are to be locked? In other words, it would be possible to define lock-check method like this?
```java
    public static boolean isEnabledFor(KeycloakSession session, Class<?> clazz) {
        Object o = session.getAttribute(ATTRIBUTE);
        return o instanceof Set && o.contains(clazz);
    }
```

Reply:
Yes, this was exactly the idea I had in mind.

Reply:
My understanding was that all entities returned from callable should be locked (in case the implementation supports locking for particular area). So from my POV it seems correct that we indicate we'd like to lock e.g. user sessions returned by `session.sessions().getUserSessionStream` or e.g. clients (`session.clients().getClientsStream`) on the logical level. And then on specific implementation we decide whether we actually lock or not (by `lockingSupportedForEntity()`).

Reply:
> My understanding was that all entities returned from callable should be locked

The current implementation will lock all entities that are loaded from the DB while the Computable is being executed even if those are not returned, so it would lock also the Clients if those are retrieved as well as a side effect if there wouldn't be the implicit `lockingSupportedForEntity` that would exclude clients. As locking of Clients happend when I tested it, the `lockingSupportedForEntity` was my first attempt to solve this.

This implicit behavior of `lockingSupportedForEntity` is difficult to understand for the caller, and might haunt us once we want to lock a Client. Once we capture the caller's intent with lockUserSessionsForModifications, such a check would not be necessary any more.

As this might need to revisit if all tests are still green, I'd analyze this in a separate ticket.

Reply:
> Do I get it right that we'd like to specify which objects could be locked on logical level?

@vramik - Yes, on a logical level.

Reply:
+1 for this approach. There will be few tweaks needed which might be tackled in the issue (specifically logical vs physical entities). @ahus1 Could you please file a Phase 1 issue for that?

Reply:
@ahus1 thank you! It makes sense to me. Please paste a link to the separate issue here once it's created, thank you.

Reply:
Done: #12717